src/Grpc.AspNetCore.Web/Internal/Base64PipeWriter.cs

```diff
@@ -0,0 +1,190 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Buffers;
+using System.Buffers.Text;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Grpc.AspNetCore.Web.Internal
+{
+    /// <summary>
+    /// Writes bytes as base64 encoded to the inner writer.
+    /// </summary>
+    internal class Base64PipeWriter : PipeWriter
+    {
+        private readonly PipeWriter _inner;
+        // We have to write original data to buffer. GetSpan/GetMemory sometimes won't return the
+        // original data if we call it again on Advance so we can't use it as temporary buffer.
+        private byte[]? _buffer;
+        private int _remainder;
+
+        // Internal for unit testing
+        internal byte _remainderByte0;
+        internal byte _remainderByte1;
+
+        public Base64PipeWriter(PipeWriter inner)
+        {
+            _inner = inner;
+        }
+
+        public override void Advance(int bytes)
+        {
+            if (bytes == 0)
+            {
+                return;
+            }
+
+            var resolvedBytes = bytes + _remainder;
+            var newRemainder = resolvedBytes % 3;
+            var bytesToProcess = resolvedBytes - newRemainder;
+
+            if (bytesToProcess > 0)
+            {
+                PreserveRemainder(_buffer.AsSpan(bytesToProcess, newRemainder));
+
+                // When writing base64 content we don't want any padding until the end.
```

Comment:
I agree with this approach though technically, in the case of gRPC-web, I believe we are allowed to add padding in the middle right?

Reply:
FlushAsync is called at the end of each message. I'll mention that in a comment.